### ISSUE TITLE
match_single_binding suggests invalid  let x = let y = ... code 

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#5267]
original code:
````rust
        let fn_sig = match self.normalize_associated_types_in_as_infer_ok(span, &fn_sig) {
            InferOk { value, obligations: o } => {
                obligations.extend(o);
                value
            }
        };
````

clippy suggested:
````
warning: this match could be written as a `let` statement
   --> src/librustc_typeck/check/method/mod.rs:370:22
    |
370 |           let fn_sig = match self.normalize_associated_types_in_as_infer_ok(span, &fn_sig) {
    |  ______________________^
371 | |             InferOk { value, obligations: o } => {
372 | |                 obligations.extend(o);
373 | |                 value
374 | |             }
375 | |         };
    | |_________^
    |
    = note: `#[warn(clippy::match_single_binding)]` on by default
````

````rust
        let fn_sig = let InferOk { value, obligations: o } = self.normalize_associated_types_in_as_infer_ok(span, &fn_sig);
        {
            obligations.extend(o);
            value
        };
````

which does not compile (yet?):

````
error[E0658]: `let` expressions in this position are experimental
   --> src/librustc_typeck/check/method/mod.rs:371:22
    |
371 |         let fn_sig = let InferOk { value, obligations: o } = self.normalize_associated_types_in_as_infer_ok(span, &fn_sig);
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: for more information, see https://github.com/rust-lang/rust/issues/53667
    = help: add `#![feature(let_chains)]` to the crate attributes to enable
````

**Comment by flip1995** [comment]
Uh, the suggestion needs to change, if the match is in a `StmtKind::Local`. The correct suggestion would be:

```rust
    let InferOk { value, obligations: o } = normalize();
    let fn_sig = {
        obligations.extend(o);
        value
    };
```

**Comment by flip1995** [comment]
cc @ThibsG

**Comment by ThibsG** [comment]
Yes, I will provide a fix for this.
